### PR TITLE
Add a `withRest` method to object cleaners

### PR DESCRIPTION
**Pull request by swansontec**
This is an alternative to #10.

This version copies all the object properties over, and then does the normal shape-aware cleaning.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -165,39 +165,56 @@ Compound cleaners don't clean data directly, but they _create_ cleaners that can
...
+  ...asThing.shape
 })
 ```
 
+In addition to the `shape` property, the returned cleaner will have a `withRest` method, which does the same thing as the cleaner, but also preserves unknown properties:
+
+```js
+// Returns `{ key: 'string', extra: false }`,
+// even though `asThing.shape` doesn't have an "extra" property:
+const y = asThing({ key: 'string', extra: false })
```
Comment (samholmes) [inline README.md]: Shouldn't this be `asThing.withRest({ key: 'string', extra: false })`?

File: src/index.js [hunk 2]
```diff
@@ -59,69 +59,97 @@ export function asArray(cleaner) {
...
 /**
- * Makes a cleaner that accepts an object with arbitrary keys
- * and the given value type. Removes keys named `__proto__` for safety.
+ * Makes a cleaner that accepts an object.
+ *
+ * If asObject receives a single cleaner function,
+ * it will will use that to clean all the object's own enumerable properties
+ * (except "__proto__", which gets filtered out).
+ *
+ * Otherwwise, if asObject receives a an object with cleaners as properties,
```
Comment (samholmes) [inline src/index.js]: Typo `receives a an object`

File: src/index.js [hunk 3]
```diff
@@ -59,69 +59,97 @@ export function asArray(cleaner) {
...
       var out = {}
-      for (var i = 0; i < keys.length; ++i) {
-        var key = keys[i]
-        if (key === '__proto__') continue
-        out[key] = cleaner(raw[key])
+      var keys = Object.keys(raw)
+      try {
+        for (var i = 0; i < keys.length; ++i) {
+          var key = keys[i]
+          if (key === '__proto__') continue
```
Comment (samholmes) [inline src/index.js:97]: When does `Object.keys` include `'__proto__'` in it's output? After checking this, it looks like Object.keys will not include properties `p` where `.hasOwnProperty(p)` returns 'false'.

Comment (swansontec) [reply src/index.js:97]: You can get this from `JSON.parse('{ "__proto__": 1 }')`. It's actually a huge source of security vulnerabilities in the ecosystem, so we eliminate that as part of cleaning.

File: src/index.js [hunk 4]
```diff
@@ -78,7 +78,7 @@ export function asArray(cleaner) {
  * it will will use that to clean all the object's own enumerable properties
  * (except "__proto__", which gets filtered out).
  *
- * Otherwwise, if asObject receives a an object with cleaners as properties,
+ * Otherwwise, if asObject receives an object with cleaners as properties,
```
Comment (samholmes) [inline src/index.js]: I just now noticed `Otherwwise` as a typo as well. >.<

**Review by samholmes — APPROVED** [review]

**Comment by samholmes** [comment]
Approved, typos aside.

**Pull request merged**